pkg/diagnostics/networkpod/service.go

```diff
@@ -0,0 +1,186 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	kcontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	kexec "k8s.io/kubernetes/pkg/util/exec"
+
+	osclient "github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/diagnostics/networkpod/util"
+	"github.com/openshift/origin/pkg/diagnostics/types"
+	sdnplugin "github.com/openshift/origin/pkg/sdn/plugin"
+)
+
+const (
+	CheckServiceNetworkName = "CheckServiceNetwork"
+)
+
+// CheckServiceNetwork is a Diagnostic to check communication between services in the cluster.
+type CheckServiceNetwork struct {
+	KubeClient *kclient.Client
+	OSClient   *osclient.Client
+
+	vnidMap map[string]uint32
+	res     types.DiagnosticResult
+}
+
+// Name is part of the Diagnostic interface and just returns name.
+func (d CheckServiceNetwork) Name() string {
+	return CheckServiceNetworkName
+}
+
+// Description is part of the Diagnostic interface and just returns the diagnostic description.
+func (d CheckServiceNetwork) Description() string {
+	return "Check pod to service communication in the cluster. In case of ovs-subnet network plugin, all pods should be able to communicate with all services and in case of multitenant network plugin, services in non-global projects should be isolated and pods in global projects should be able to access any service in the cluster."
+}
+
+// CanRun is part of the Diagnostic interface; it determines if the conditions are right to run this diagnostic.
+func (d CheckServiceNetwork) CanRun() (bool, error) {
+	if d.KubeClient == nil {
+		return false, errors.New("must have kube client")
+	} else if d.OSClient == nil {
+		return false, errors.New("must have openshift client")
+	}
+	return true, nil
+}
+
+// Check is part of the Diagnostic interface; it runs the actual diagnostic logic
+func (d CheckServiceNetwork) Check() types.DiagnosticResult {
+	d.res = types.NewDiagnosticResult(CheckServiceNetworkName)
+
+	pluginName, ok, err := util.GetOpenShiftNetworkPlugin(d.OSClient)
+	if err != nil {
+		d.res.Error("DSvcNet1001", err, fmt.Sprintf("Checking network plugin failed. Error: %s", err))
+		return d.res
+	}
+	if !ok {
+		d.res.Warn("DSvcNet1002", nil, fmt.Sprintf("Skipping service connectivity test. Reason: Not using openshift network plugin."))
+		return d.res
+	}
+
+	services, err := getAllServices(d.KubeClient)
+	if err != nil {
+		d.res.Error("DSvcNet1003", err, fmt.Sprintf("Getting all services failed. Error: %s", err))
+		return d.res
+	}
+	if len(services) == 0 {
+		d.res.Warn("DSvcNet1004", nil, fmt.Sprintf("Skipping service connectivity test. Reason: No services found."))
```

Comment:
If it's a warning, what should the user do about this? I would think having no services would be a very strange condition.